notifications/v1/client.go

```diff
@@ -0,0 +1,95 @@
+package v1
+
+import (
+	"context"
+	"encoding/json"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/sourcegraph/log"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// Client is a client for receiving SAMS notifications from a GCP Pub/Sub
+// subscription.
+type Client struct {
+	logger        log.Logger
+	subscription  *pubsub.Subscription
+	cancelContext context.CancelFunc
+}
+
+// NewClient creates a new Client for receiving SAMS notifications with given
+// GCP project ID and Pub/Sub subscription ID.
+func NewClient(logger log.Logger, projectID, subscriptionID string) (*Client, error) {
+	client, err := pubsub.NewClient(context.Background(), projectID)
+	if err != nil {
+		return nil, errors.Wrap(err, "create GCP Pub/Sub client")
+	}
+	return &Client{
+		logger:       logger.Scoped("notification.ClientV1"),
+		subscription: client.Subscription(subscriptionID),
+	}, nil
+}
+
+// ReceiveHandlers is a collection of receive handlers for each type of SAMS
+// notifications. If the handler of a notification is nil, the notification will
+// be acknowledged automatically without any processing.
+//
+// If a handler returns an error, the notification will be unacknowledged and
+// retried later.
+type ReceiveHandlers struct {
+	// OnUserDeleted is called when a "UserDeleted" notification is received.
+	OnUserDeleted func(data *UserDeletedData) error
+}
+
+// Receive starts receiving SAMS notifications and calls the corresponding
+// handler for each notification. It blocks until the context is done (e.g.
+// deadline exceed or canceled).
+//
+// ⚠️ WARNING: Each subscription can only have one active receiver at a time,
+// i.e. there should only be one call to Receive for a given subscription.
+func (c *Client) Receive(handler *ReceiveHandlers) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	c.cancelContext = cancel
+	return c.subscription.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
```

Comment:
Can we receive messages concurrently, or is it only ever 1 at a time? Do we want to support concurrent handling, or is that a handler implementation details? Should handlers expect to be concurrency-safe?

Reply:
Good call on OOTB concurrency control (currently 1 at a time)! But I think whether or not handlers can be concurrency-safe is defined by each handler implementation (i.e. do they do it in a concurrency-safe manner).

Reply:
Actually, re-read the `subscription.Receive` docstring, it says

>Receive calls f concurrently from multiple goroutines.

but it does provide an option to sort of limiting the max concurrency.

Reply:
I decided to expose the `pubsub.ReceiveSettings` in https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24/commits/08bb5832ea6edd81e434500f7b49d4b9dabcce40, PTAL!

Reply:
Depends on outcome of https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24#discussion_r1594295009

Reply:
LGTM, though we may want to make some changes depending on discussion in https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24#discussion_r1594295009